lib/AST/ASTMangler.cpp

```diff
@@ -2506,17 +2526,47 @@ void ASTMangler::appendProtocolName(const ProtocolDecl *protocol,
     appendDeclName(protocol);
 }
 
-const clang::NamedDecl *ASTMangler::getClangDeclForMangling(const ValueDecl *vd) {
-  auto namedDecl =  dyn_cast_or_null<clang::NamedDecl>(vd->getClangDecl());
+const ASTMangler::ClangDeclForManglingContainer
+ASTMangler::getClangDeclForMangling(const ValueDecl *vd) {
+  auto namedDecl = dyn_cast_or_null<clang::NamedDecl>(vd->getClangDecl());
   if (!namedDecl)
     return nullptr;
-  
+
   // Use an anonymous enum's enclosing typedef for the mangled name, if
   // present. This matches C++'s rules for linkage names of tag declarations.
   if (namedDecl->getDeclName().isEmpty())
-    if (auto *tagDecl = dyn_cast<clang::TagDecl>(namedDecl))
-      if (auto *typedefDecl = tagDecl->getTypedefNameForAnonDecl())
```

Comment:
This is a little weird. I want to make the Clang AST look like the Objective-C definition so that the mangled name will be the same, but I don't want the mutation to persist beyond after the mangler is done with it. More specifically, I want the anonymous enum to no longer be anonymous and instead have the backing typedef's name.

I considered making a copy of the node, but that seems pretty wasteful. I'd have to copy at least once per NS_OPTION use and keep them around somewhere. Instead I've made this RAII container that just clears the anonymous enum's name after the client is done with it.